packages/react-reconciler/src/ReactFiberWorkLoop.new.js

```diff
@@ -716,21 +716,27 @@ function ensureRootIsScheduled(root: FiberRoot, currentTime: number) {
     // Special case: There's nothing to work on.
     if (existingCallbackNode !== null) {
       cancelCallback(existingCallbackNode);
-      root.callbackNode = null;
-      root.callbackPriority = NoLanePriority;
     }
+    root.callbackNode = null;
+    root.callbackPriority = NoLanePriority;
     return;
   }
 
   // Check if there's an existing task. We may be able to reuse it.
-  if (existingCallbackNode !== null) {
-    const existingCallbackPriority = root.callbackPriority;
-    if (existingCallbackPriority === newCallbackPriority) {
-      // The priority hasn't changed. We can reuse the existing task. Exit.
-      return;
+  const existingCallbackPriority = root.callbackPriority;
+  if (existingCallbackPriority === newCallbackPriority) {
+    if (__DEV__) {
+      invariant(
+        existingCallbackNode,
+        'Expected scheduled callback to exist. This error is likely caused by a bug in React. Please file an issue.',
+      );
     }
-    // The priority changed. Cancel the existing callback. We'll schedule a new
-    // one below.
+    // The priority hasn't changed. We can reuse the existing task. Exit.
+    return;
+  }
+
+  if (existingCallbackNode != null) {
+    // Cancel the existing callback. We'll schedule a new one below.
```

Comment:
I think the warning you wanted goes in the else block of this check.

```js
if (existingCallbackNode != null) {
  // Cancel the existing callback. We'll schedule a new one below.
  cancelCallback(existingCallbackNode);
} else {
  if (__DEV__ && newCallbackPriority !== InputDiscreteLanePriority && newCallbackPriority !==  SyncLanePriority) {
    console.error('oh noes!');
  }
} 
```

Reply:
If it goes in this block, then we may have already returned without something being scheduled.

Reply:
Also, there will be no callback on the first schedule.

Reply:
One more nit: generally, for our own internal types, we always explicitly compare to `null`. That way 1) a missing field/argument is distinct from one that is intentionally null, and 2) it's a tiny bit faster.

Reply:
> If it goes in this block, then we may have already returned without something being scheduled.

Can you rephrase? The check I'm suggesting you move is already behind a __DEV__  check which means it doesn't affect behavior. If it did then we'd definitely want to remove the __DEV__ check :D

Reply:
I thought so, but then there were a bunch of invariant calls that didn't do a null check so I switched it.

Fixed 👍

Reply:
Yeah, if we move the check from the block above this here, and the control flow returns in the above block, then it will never hit this invariant check. Which means we would chose to reuse and exiting task that doesn't exist.

Reply:
Ok you can leave that there then but please make it a warning, not an error.

I was thinking you wanted to warn about a slightly different case, which is if you expect to cancel a non-microtask task but the callback node doesn't exist as expected.